### PR TITLE
[Operator] Type promotion for pointwise Ops

### PR DESCRIPTION
**Pull request by Bowen12992**
- Add Type promotion for pointwise Ops 
- Add trunc & floor mode for divOp
- Move Type promotion into pointwise JIT 
- Support different condition of whereOp
- Add UTs of type promotion for pointwise Ops
- Fix related docs

### CODE REVIEW

File: src/flag_gems/utils/type_utils.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import torch._prims_common as utils
+
+
+def type_promotion(*args, type_promotion: utils.ELEMENTWISE_TYPE_PROMOTION_KIND):
```
Comment (iclementine) [inline src/flag_gems/utils/type_utils.py:4]: Maybe we can use a simpler implementation for type promotion rules.(e.g. use `torch.promote_types` or `torch.result_type` for `ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT`, and write our own implementation for other cases).

File: src/flag_gems/utils/pointwise_dynamic.py [hunk 2]
```diff
@@ -50,12 +51,18 @@ def __init__(
         is_tensor: Optional[List[bool]] = None,
         dtypes: Optional[List[Optional[type]]] = None,
         num_outputs: Optional[int] = None,
-        output_dtypes: Optional[List[torch.dtype]] = None,
+        promotion_methods: Optional[List[Tuple[int, ...]]] = None,
```
Comment (iclementine) [inline src/flag_gems/utils/pointwise_dynamic.py:54]: What about still support `List[torch.dtype]`?

Comment (Bowen12992) [reply src/flag_gems/utils/pointwise_dynamic.py:54]: I thought because `pointwise_dynamic` is for element_wise ops and once we add type promotion it can deal with more complicated input type combination，so we should use promotion to get all kind output type for different type combination. If we still use `List[torch.dtype]` to specify output，on the one hand，it‘s hard to figure out all kind output type for different type combination, on the other hand, it makes  the logic of `pointwise_dynamic` much more complicated and confusing

File: src/flag_gems/utils/type_utils.py (hunk 1 above)
Comment (Bowen12992) [reply src/flag_gems/utils/type_utils.py:4]: Yeah, we can simplify the logic!  but I have two concerns. 1. Will Gems support dynamic-shape with maybe `sympy` in the feature？2. I‘m going to figure out  if there are differences among `torch.promote_types`  `torch.result_type` and  `ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT`

File: src/flag_gems/ops/div.py [hunk 3]
```diff
@@ -6,32 +6,106 @@
...
+    elif isinstance(A, torch.Tensor):
+        return floor_div_func_tensor_scalar(A, B)
+    elif isinstance(B, torch.Tensor):
+        return floor_div_func_scalar_tensor(A, B)
+    else:
+        # Both scalar
+        return A // B
+
+
+def div(A, B, rounding_mode=None):
```
Comment (StrongSpoon) [inline src/flag_gems/ops/div.py:102]: aten interface div.Tensor doesn't have argument named rounding_mode. do you want to implement 'div.Tensor_mode'?

Comment (StrongSpoon) [reply src/flag_gems/ops/div.py:102]: besides, trunc_divide might come to accuracy error due to low precision, and we didn't solve it last time.

Comment (Bowen12992) [reply src/flag_gems/ops/div.py:102]: I use  `div_round_zero` for  `trunc_divide` rather than `O_val = tl.math.trunc([X_val.to](http://x_val.to/)(tl.float64) / Y_val)` here and did some tests, i didn't get accuracy error due to low precision any more

File: src/flag_gems/utils/type_utils.py (hunk 1 above)
Comment (iclementine) [reply src/flag_gems/utils/type_utils.py:4]: We may not support symbolic shape since flaggems is basically designed for processing real tensors, with concrete shapes.

**Review by iclementine — APPROVED** [review]
LGTM
